src/maybe_uninit.rs

```diff
@@ -0,0 +1,105 @@
+
+
+use array::Array;
+use std::mem::ManuallyDrop;
+use std::mem::uninitialized;
+
+/// A combination of ManuallyDrop and “maybe uninitialized”;
+/// this wraps a value that can be wholly or partially uninitialized;
+/// it also has no drop regardless of the type of T.
+///
+/// This is a stop-gap solution until MaybeUninit is stable in Rust's std.
+#[repr(C)]
+pub struct MaybeUninit<T>(ManuallyDrop<T>);
+
+impl<T> MaybeUninit<T> {
+    /// Create a new MaybeUninit with uninitialized interior
+    pub unsafe fn uninitialized() -> Self {
+        Self::from(uninitialized())
```

Comment:
I'm afraid this is not correct -- it is as incorrect as just plain `mem::uninitialized`.

Using terminology from [my blog post](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html), the issue with `mem::uninitialized` is that it returns data that violates the validity invariant. While `ManuallyDrop` can handle data that is already dropped, it does not have any effect on the validity invariant: `ManuallyDrop<bool>` may only be `0x0` or `0x1`, and hence 
```rust
let x: MaybeUninit<bool> = MaybeUninit::uninitialized();
```
with your type is UB.

For non-`Copy` types, there is currently no way to do this. We should really work towards solving the remaining open questions in https://github.com/rust-lang/rust/issues/53491!

Reply:
FWIW; I know this is not correct, this is the current best effort with stable Rust like the PR says.

The current state of arrayvec as it is released, can't be said to be any better than this, I'm sure?

Reply:
@RalfJung With MaybeUninit, what's the importance of using `MaybeUninit<[T; 10]>` vs `[MaybeUninit<T>; 10]` ? The latter I don't even have clear picture of how to initialize, when we start with `T` values that are not `Copy`.

Reply:
> The current state of arrayvec as it is released, can't be said to be any better than this, I'm sure?

I do not know that state. This is pretty much equivalent to using `mem::uninitialized` directly.

> it uses mem::uninitialized() to initialize a whole value of [MaybeUninit<T>; 10].

oO I had no idea.^^ Where can I find its sources?

Time that we get `mem::uninitialized` deprecated so that these things are rejected by `-D warnings`.

But given that `MaybeUninit` is a union, it is actually okay to not initialize it. The problem with `mem::uninitialized` is that it almost always violates the validity invariant of the type it returns; but for `MaybeUninit`, the validity invariant accepts literally any data (including uninitialized data), and hence there is no problem with that particular usage of `mem::uninitialized`.

Reply:
Looking at `ArrayVec` when it got removed in rust-lang/rust@687cc292fd681be9739dc973acd5eaa5f73a5ce7, it uses `mem::uninitialized` on `ManuallyDrop`, which is *not* okay.

Reply:
Ok! By the way I'm following the stacked borrows discussion and unsafe code guidelines now. It will have a big impact on ndarray, I believe. Just like the rest of the ecosystem we are gradually getting a clearer picture of what we can do in unsafe code and can dot our i's and cross the t's here.

I would go with using `MaybeUninit<[T; 10]>` and offsetting into it, because the other way around leads to code with more transmutes when getting initialized `[T; 10]` wrapped into the ArrayVec.

Given this, it sounds like it is ok to use `mem::uninitialized` with `[MaybeUninit<T>; 10]`, even if the array is the outer type. That's the only answer to how to set a value to such a field that I can see (or equivalent union trick).